### PR TITLE
Update to gzip encode/decode and bounds checking

### PR DESCRIPTION
**Pull request by stevenblair**
- adjusted threshold for gzip activation
- added bounds checking on quality decode
- improved readme and added full example code

### CODE REVIEW

**Comment by jsnctl** [comment]
Very cool

image

File: decoder.go [hunk 1]
```diff
@@ -282,8 +284,11 @@ func (s *Decoder) DecodeToBuffer(buf []byte, totalLength int) error {
 				sampleNumber = actualSamples
 			} else {
 				// write up to valUnsigned remaining Q values for this variable
-				for j := sampleNumber + 1; j < int(valUnsigned); j++ {
-					s.Out[j].Q[i] = s.Out[sampleNumber].Q[i]
+				for j := sampleNumber + 1; j < int(valUnsigned); /*&& j < sampleNumber*/ j++ {
+					if sampleNumber < len(s.Out) && j < len(s.Out) {
+						// fmt.Println(i, j, valUnsigned, sampleNumber)
+						s.Out[j].Q[i] = s.Out[sampleNumber].Q[i]
+					}
```
Comment (jsnctl) [inline decoder.go]: Out of interest: what was the behaviour of this prior to adding the conditional?

Comment (jsnctl) [reply decoder.go]: And is there a way we can add a test for the change? We're not really doing `Encoder`/`Decoder`-level unit tests, so might not be appropriate (or even easy to test in the first place), but thought I'd ask anyway

**Review by jsnctl — APPROVED** [review]
Just a single Q about the implementation change (mostly out of interest), and maybe a thought on testing for future consideration

File: decoder.go (hunk 1 above)
Comment (stevenblair) [reply decoder.go]: This happens when fewer samples are encoded than the configured value (i.e. encoding stops part of the way through a message). There is a test for this scenario, but it doesn't cover a specific combination of inputs which could cause an array overrun - which I think can only happen if the encoder inserts invalid data. This change protects against that.

In general, there's definitely scope for further testing of invalid inputs.

**Pull request merged**